### PR TITLE
Verify NFT content media type before pinning

### PR DESCRIPTION
**Pull request by cypt4**
Resolves https://github.com/brave/brave-browser/issues/28775

Resolves 

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
After an NFT added to the wallet, media type of the "image" field is checked. So if it differs from "image/.." NFT won't be pinned. Probably we can test this by sniffing NFT gateway for the https://nftstorage.link(example)/ipfs/IMAGE_CID request by changing content_type here from image/png to blablabla

### CODE REVIEW

File: components/brave_wallet/common/brave_wallet.mojom [hunk 1]
```diff
@@ -222,7 +222,8 @@ enum WalletPinServiceErrorCode {
   ERR_WRONG_METADATA_FORMAT = 4,
   ERR_ALREADY_PINNED = 5,
   ERR_NOT_PINNED = 6,
-  ERR_PINNING_FAILED = 7
+  ERR_PINNING_FAILED = 7,
+  ERR_MEDIA_TYPE_UNSUPORTED = 8
```
Comment (supermassive) [inline components/brave_wallet/common/brave_wallet.mojom:226]:
```
UNSUPORTED ->
UNSUPPORTED
```

File: components/brave_wallet/browser/brave_wallet_pin_service.cc [hunk 2]
```diff
@@ -149,11 +152,94 @@ std::string BraveWalletPinService::ErrorCodeToString(
       return "ERR_NOT_PINNED";
     case mojom::WalletPinServiceErrorCode::ERR_PINNING_FAILED:
       return "ERR_PINNING_FAILED";
+    case mojom::WalletPinServiceErrorCode::ERR_MEDIA_TYPE_UNSUPORTED:
+      return "ERR_MEDIA_TYPE_UNSUPORTED";
```
Comment (supermassive) [inline components/brave_wallet/browser/brave_wallet_pin_service.cc:156]: UNSUPPORTED

File: components/brave_wallet/browser/brave_wallet_pin_service.cc [hunk 3]
```diff
@@ -507,20 +595,56 @@ void BraveWalletPinService::OnTokenMetaDataReceived(
...
+    auto image_cid = ExtractCID(*image_url);
     if (image_cid) {
-      cids.push_back(image_cid.value());
+      cids.push_back(*image_cid);
+      content_type_checker_->CheckContentTypeSupported(
+          *image_url,
+          base::BindOnce(&BraveWalletPinService::OnContentTypeChecked,
+                         weak_ptr_factory_.GetWeakPtr(), std::move(service),
+                         std::move(token), std::move(cids),
+                         std::move(callback)));
```
Comment (supermassive) [inline components/brave_wallet/browser/brave_wallet_pin_service.cc:608]: This needs S&P review as we are getting arbitrary url resource from blockchain and forcing browser to make a request to it.

File: components/brave_wallet/browser/brave_wallet_pin_service.h [hunk 4]
```diff
@@ -23,10 +23,42 @@
...
+ private:
+  using UrlLoaderList = std::list<std::unique_ptr<network::SimpleURLLoader>>;
+
+  void OnHeadersFetched(UrlLoaderList::iterator iterator,
+                        base::OnceCallback<void(absl::optional<bool>)> callback,
+                        scoped_refptr<net::HttpResponseHeaders> headers);
+
+  raw_ptr<PrefService> pref_service_;
+  scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory_;
+  // Reports that are actively being sent.
```
Comment (supermassive) [inline components/brave_wallet/browser/brave_wallet_pin_service.h:56]: What reports?

File: components/brave_wallet/browser/brave_wallet_pin_service.cc (hunk 3 above)
Comment (cypt4) [reply components/brave_wallet/browser/brave_wallet_pin_service.cc:608]: It is a bit different : actually we make a call to the IPFS gateway which we use to fetch metadata on the previous step. Final url looks like https://ipfs.io/CID/...

Comment (stoletheminerals) [reply components/brave_wallet/browser/brave_wallet_pin_service.cc:611]: I think it's fine. It's not arbitrary, like @cypt4 said and we only call IPFS gateway. This code also doesn't fetch a file, only response headers.

**Review by supermassive — APPROVED** [review]
👍

**Review by stoletheminerals — APPROVED** [review]

**Comment by srirambv** [comment]
Verification passed on 

Brave | 1.51.32 Chromium: 111.0.5563.64 (Official Build) nightly (32-bit)
-- | --
Revision | `c710e93d5b63b7095afe8c2c17df34408078439d-refs/branch-heads/5563@{#995}`
OS | Windows 11 Version 22H2 (Build 23403.1001)

- Verified steps from PR
- Verified when `content-type` is not `image/png` NFT Pinning fails 
image

**Pull request merged**